### PR TITLE
Impro 174 psychrometric calculations wet bulb

### PR DESCRIPTION
**Pull request by wxtim**
Issue: #194 

Testing:
 - [ ] Ran tests and they passed OK
 - [ ] Added new tests for the new feature(s)

CLA
 - [ ] If a new developer, signed up to CLA

### CODE REVIEW

File: lib/improver/constants.py [hunk 1]
```diff
@@ -41,3 +41,6 @@
 
 # Specific heat capacity of dry air (J K-1 kg-1)
 CP_DRY_AIR = 1005.0
+
+# Triple Point of Water
```
[the remark this answers is not in the document]
Comment (wxtim) [reply lib/improver/constants.py]: Yes. Done.

File: lib/improver/tests/psychrometric_calculations/test_psychrometric_calculations.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    cube.units = Unit(units)
+    return cube
+
+
+class Test_check_range(IrisTest):
+
+    """Tests the check_range method
+    """
+
+    def test_fail(self):
```
[the remark this answers is not in the document]
Comment (wxtim) [reply lib/improver/tests/psychrometric_calculations/test_psychrometric_calculations.py]: Done

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+                "Highest temperature = {}")
+        raise TypeError(emsg.format(low,
+                                    high,
+                                    cube.data.max(),
+                                    cube.data.min()))
+
+
+def saturation_vapour_pressure_ashrae(temperature):
+    ''' Function to compute saturation vapour pressure in [kPa]
+        ASHRAE Fundamentals handbook (2005) p 6.2, equation 5 and 6.
```
[the remark this answers is not in the document]
Comment (wxtim) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: This hath now been done. Will push shortly

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+    Parameters
```
[the remark this answers is not in the document]
Comment (wxtim) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Thank you. I am sorting this.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+                                cell + constant_4 * cell ** 2 + constant_5 *
+                                cell ** 3 + constant_6 * cell ** 4 +
+                                constant_7 * np.log(cell))) / 1000
+        else:
+            cell[...] = (np.exp(constant_8 / cell + constant_9 + constant_10 *
+                                cell + constant_11 * cell ** 2 + constant_12 *
+                                cell ** 3 + constant_13 * np.log(cell))) / 1000
+    result = temp.copy(data=data)
+    result.units = Unit('kPa')
+    result.convert_units('Pa')
```
[the remark this answers is not in the document]
Comment (wxtim) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: yes. Done

File: lib/improver/tests/psychrometric_calculations/test_psychrometric_calculations.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        data = np.array([[5., 290.],
+                         [290., 290.]])
+        cube_in = _make_test_cube("temperature", "K", data=data)
+        with self.assertRaisesRegexp(TypeError, emsg):
+            check_range(cube_in, 10., 300.)
+
+
+class Test_calculate_svp_ashrae(IrisTest):
+
+    """saturation_vapour_press == svp"""
```
[the remark this answers is not in the document]
Comment (wxtim) [reply lib/improver/tests/psychrometric_calculations/test_psychrometric_calculations.py]: I have responded to this.

File: lib/improver/tests/psychrometric_calculations/test_psychrometric_calculations.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        emsg = "only valid for temperatures between"
+        with self.assertRaisesRegexp(ValueError, emsg):
+            check_range(cube_in, 10., 300.)
+        data = np.array([[5., 290.],
+                         [290., 290.]])
+        cube_in = _make_test_cube("temperature", "K", data=data)
+        with self.assertRaisesRegexp(ValueError, emsg):
+            check_range(cube_in, 10., 300.)
+
+    def test_succeeds(self):
```
[the remark this answers is not in the document]
Comment (wxtim) [reply lib/improver/tests/psychrometric_calculations/test_psychrometric_calculations.py]: Addressed

File: lib/improver/psychrometric_calculations/psychrometric_plugins.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            emsg = "Temperature is not a cube, but {}"
+            raise TypeError(emsg.format(type(temperature)))
+        if not isinstance(rel_humidity, iris.cube.Cube):
+            emsg = "Temperature is not a cube, but {}"
+            raise TypeError(emsg.format(type(rel_humidity)))
+        if not isinstance(pressure, iris.cube.Cube):
+            emsg = "Pressure is not a cube, but {}"
+            raise TypeError(emsg.format(type(pressure)))
+
+        # Check that cubes are all the same shape
```
Comment (wxtim) [inline lib/improver/psychrometric_calculations/psychrometric_plugins.py]: Might want to use utility fn here? Besides if this is wrong you will get a sensible broadcast/shape error from IRIS, so is it needed at all?

File: lib/improver/psychrometric_calculations/psychrometric_plugins.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+        Returns:
+            wet_bulb : iris.cube.Cube
+                Cube containing the wet bulb temperature
+
+        Raises:
+            TypeError: If any of the inputs are not cubes.
+            ValueError: If any of the input cubes are not the same shape.
+        """
+        # Check that inputs are of correct type
```
Comment (wxtim) [inline lib/improver/psychrometric_calculations/psychrometric_plugins.py]: Necessary?
Should we check height coords match?

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module to contain Psychrometric Calculations."""
+
+from cf_units import Unit
+import improver.constants as constant
+import iris.analysis.maths as maths
+import numpy as np
+
+
+def check_range(cube, low, high):
```
Comment (wxtim) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: This should probably be in utilities shouldn't it?

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+    Args:
+        temperature : iris.cube.Cube
+            Cube of temperature which will be converted to Celsius
+            prior to calculation
+        rel_humidity : iris.cube.Cube
+            Cube of relative humidity
+        pressure : iris.cube.Cube or float
+            Cube of pressure which will be converted to kilopPascals
+            prior to calculation
+        precision : float
```
Comment (wxtim) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: p_units and precision should be kwargs

File: lib/improver/tests/psychrometric_calculations/test_psychrometric_calculations.py [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+    pressure using the ASHRAE function"""
+    def test_basic(self):
+        """test to check that the saturation_vapour_pressure
+        method returns a cube with answers calculated to be correct
+        """
+        temp = _make_test_cube("temperature", "K")
+        expected_data = np.array([[470.08, 470.08], [1919.60, 1919.60]])
+        result = saturation_vapour_pressure_ashrae(temp)
+        self.assertArrayAlmostEqual(result.data, expected_data, decimal=2)
+        self.assertEqual(result.units, Unit('Pa'))
```
Comment (wxtim) [inline lib/improver/tests/psychrometric_calculations/test_psychrometric_calculations.py]: Check result is a cube!

File: lib/improver/tests/psychrometric_calculations/test_psychrometric_calculations.py [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+
+    def test_edge_conditions(self):
+        """The wet bulb algorithm is not stable for all possible combinations
+        of values. This test looks at a set of meteorlogical edge case
+        scenarios to check that the method works.
+        Lowest Recorded RH = 0.001
+        Recorded temperature range -95 ~< T ~< 51 Celsius
+        Pressure Max  ~ 1085 hPa
+        """
+        scenarios = [{'t': -100., 'rh': 0.0001, 'p': .1},
```
Comment (wxtim) [inline lib/improver/tests/psychrometric_calculations/test_psychrometric_calculations.py]: Consider adding desired values
Consider adding a high temp moderate pressure

File: lib/improver/psychrometric_calculations/psychrometric_plugins.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+from improver.psychrometric_calculations.psychrometric_calculations import (
+    wet_bulb)
+
+
+class WetBulb(object):
+    """Calculate the wet bulb temperature given cubes of temperature,
+    relative humidity and pressure"""
+
+    def __init__(self):
```
Comment (wxtim) [inline lib/improver/psychrometric_calculations/psychrometric_plugins.py]: Does this need a test?

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py (hunk 10 above)
Comment (bayliffe) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: I'd just make it a private function. Given the specificity of the error message, it cannot be reused elsewhere.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 15]
```diff
@@ -0,0 +1,411 @@
...
+    constant_4 = 0.00000062215701
+    constant_5 = 2.0747825E-09
+    constant_6 = -9.484024E-13
+    constant_7 = 4.1635019
+    constant_8 = -5800.2206
+    constant_9 = 1.3914993
+    constant_10 = -0.048640239
+    constant_11 = 0.000041764768
+    constant_12 = -0.00000001445209310000
+    constant_13 = 6.5459673
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: This could be tidied up with a dictionary or list, e.g.:
`constants = {1: -5674.5359, 2: 6.3925247, 3: -0.009677843, ...}`
`cell[...] = (np.exp(constant[1] / cell` etc

This is my preference, rather than a requirement.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+    hr = temp.copy(data=result_numer / result_denom)
+
+    # Tidying up cube
+    hr.rename("Humidity ratio")
+    hr.units = Unit("1")
+    return hr
+
+
+def humidity_ratio_fm_wb(temperature, wet_bulb, pressure):
+    ''' Function to compute humidity ratio given a temperature pressure and
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: I miss """

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 17]
```diff
@@ -0,0 +1,411 @@
...
+    svp = saturation_vapour_pressure_goff_gratch(temp, press)
+    svp.convert_units('kPa')
+
+    # Calculation
+    result_numer = (0.62198 * rh.data * svp.data)
+    result_denom = (press.data - (rh.data * svp.data))
+    hr = temp.copy(data=result_numer / result_denom)
+
+    # Tidying up cube
+    hr.rename("Humidity ratio")
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Not a valid name, `humidity_ratio` perhaps?

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 18]
```diff
@@ -0,0 +1,411 @@
...
+    temp.convert_units('celsius')
+    rh = relative_humidity.copy()
+    rh.convert_units(1)
+    press = pressure.copy()
+    press.convert_units('kPa')
+    svp = saturation_vapour_pressure_goff_gratch(temp, press)
+    svp.convert_units('kPa')
+
+    # Calculation
+    result_numer = (0.62198 * rh.data * svp.data)
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Include a comment describing the calculation of partial pressure of water vapour (Pw), e.g. RH = Pw/Psat

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 19]
```diff
@@ -0,0 +1,411 @@
...
+    hr.units = Unit("1")
+    return hr
+
+
+def humidity_ratio_fm_wb(temperature, wet_bulb, pressure):
+    ''' Function to compute humidity ratio given a temperature pressure and
+    wet bulb temperature.
+
+    Args:
+        temp: iris.cube.Cube
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: temperature

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 20]
```diff
@@ -0,0 +1,411 @@
...
+    press = pressure.copy()
+    press.convert_units("kPa")
+    wb = wet_bulb.copy()
+    wb.convert_units("celsius")
+
+    # Calculation
+    wb_depression = temp - wb
+    svp = saturation_vapour_pressure_goff_gratch(temp, press)
+    svp.convert_units('kPa')
+    ws = 0.62198 * svp / (press - svp)
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Why are you using SVP here, not SVP*RH?

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 21]
```diff
@@ -0,0 +1,411 @@
...
+def wet_bulb(temperature, relative_humidity, pressure,
+             p_units="hPa", precision=0.00001):
+    """
+    Calculates the Wet Bulb Temperature Using Newton-Raphson iteration
+
+    Args:
+        temperature : iris.cube.Cube
+            Cube of temperature which will be converted to Celsius
+            prior to calculation
+        rel_humidity : iris.cube.Cube
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: relative_humidity

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 22]
```diff
@@ -0,0 +1,411 @@
...
+            Cube of relative humidity
+        pressure : iris.cube.Cube or float
+            Cube of pressure which will be converted to kilopPascals
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Typo: KiloPascals

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py (hunk 11 above)
Comment (bayliffe) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Agreed, there should be a default precision is the user does not provide one.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 23]
```diff
@@ -0,0 +1,411 @@
...
+        raise ValueError(emsg.format(press.data.min()))
+    if rh.data.min() == 0:
+        emsg = ("This Wet Bulb Temperature algorithm is"
+                " only valid for relative humidities greater than"
+                " Zero. Input cube has\n"
+                "Lowest rh of:  {}\n")
+        raise ValueError(emsg.format(rh.data.min()))
+    # create a numpy array of precision and cube of increment
+    precision = np.full(temp.data.shape, precision)
+    increment = temp.copy(data=np.full(rh.shape, 0.001))
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: temp.shape surely, to make sure the temp cube coordinates match?

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 24]
```diff
@@ -0,0 +1,411 @@
...
+        dw_dthr.units = Unit(1)
+        result2 = result - (hr_new - hr_normal) / dw_dthr
+        result2.units = Unit('celsius')
+        data = np.where((maths.abs((hr_new - hr_normal) /
+                                   hr_normal)).data > precision,
+                        result2.data,
+                        result.data)
+        result = result.copy(data=data)
+        result.units = Unit('celsius')
+        hr_new = humidity_ratio_fm_wb(temp, result, press)
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: This is not very lovely. I've not gone through it in detail, I'd prefer to do that when you reconsider how you've implemented your iterator. Things to consider:
* How much copying is necessary?
* Do the units need to be set at every stage?
* Are there python tools that could help clean this up (scipy.optimize)?

File: lib/improver/psychrometric_calculations/psychrometric_plugins.py (hunk 14 above)
Comment (bayliffe) [reply lib/improver/psychrometric_calculations/psychrometric_plugins.py]: No.

File: lib/improver/psychrometric_calculations/psychrometric_plugins.py (hunk 9 above)
Comment (bayliffe) [reply lib/improver/psychrometric_calculations/psychrometric_plugins.py]: I assume this will be down the processing chain and so that will have been done earlier. The below tests that you are receiving a cube seem excessive.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 25]
```diff
@@ -0,0 +1,411 @@
...
+    wb = wet_bulb.copy()
+    wb.convert_units("celsius")
+
+    # Calculation
+    wb_depression = temp - wb
+    svp = saturation_vapour_pressure_goff_gratch(temp, press)
+    svp.convert_units('kPa')
+    ws = 0.62198 * svp / (press - svp)
+    # This is inefficient because it produces both answers
+    above_zero = (((2501. - 2.326 * wb.data) * ws.data -
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: You need to indicate that these are equations 35/37 in the ASHRAE document, both here and in the docstring.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 26]
```diff
@@ -0,0 +1,411 @@
...
+    temp = temperature.copy()
+    temp.convert_units('K')
+    check_range(temp, 173., 473.)
+    data = temp.data
+    for cell in np.nditer(data, op_flags=['readwrite']):
+        if cell <= 273.15:
+            cell[...] = (np.exp(constant_1 / cell + constant_2 + constant_3 *
+                                cell + constant_4 * cell ** 2 + constant_5 *
+                                cell ** 3 + constant_6 * cell ** 4 +
+                                constant_7 * np.log(cell))) / 1000
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: (This is not actually true in Python, but it is in some languages. Best avoided I would say).

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py (hunk 24 above)
Comment (wxtim) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: I've managed to remove some of the copying and unit changing, but not by any means as much as I'd like.

Inside the while loop there is now only 1 .copy(), which I believe is necessary. There are no unit conversions inside the while loop (the units are forced where they are changed rather than converted.) except for those in the humidity_ratio_fm_wb function.

Part of this ugliness (esp around line ~407) is caused by needing to avoid
`NotYetImplementedError: Cannot use 'subtract' with differing units (celsius & kelvin)`
Sadly Iris does not really have concept of a temperature delta - which is exactly what is needed here.

I had a look at SciPy Optimize, and although there is a Newton-Raphson method, it seems not to be designed for use on arrays, and attempts to do so return an array comparison ambiguous, please use .any() or .all() error.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 27]
```diff
@@ -0,0 +1,425 @@
...
+                                     cube.data.max()))
+
+
+def saturation_vapour_pressure_ashrae(temperature):
+    ''' Compute Saturation vapour pressure from temperature, using ASHRAE
+    (American Society of Heating, Refrigerating and Air-Conditioning
+    Engineers) method.
+
+    Args:
+        temp : iris.cube.Cube
```
Comment (bayliffe) [inline lib/improver/psychrometric_calculations/psychrometric_calculations.py]: Called temperature.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py [hunk 28]
```diff
@@ -0,0 +1,411 @@
...
+    # Copy cubes to avoid modifying those passed to this function
+    # Check that units are correct for the calculations
+    press = pressure.copy()
+    press.convert_units('hPa')
+    temp = temperature.copy()
+    temp.convert_units('K')
+    check_range(temp, 173., 373.)
+
+    # create output cube
+    svp = temp.copy()
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: This is technically okay because there is a temperature.copy() in the calling routine (which currently calls goff-gratch method, but could call this). However, the routine is overwriting the passed in temperature cube data in-situ. If someone were to call this routine as a standalone function with a temperature cube, they would find their input data modified. Maybe the cube.copy() that is going on in the calling function should be moved in to each of the saturated vapour pressure functions to avoid this issue. Any thoughts?

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py (hunk 18 above)
Comment (bayliffe) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: I thought this was the basis of this calculation. Am I wrong? If so, please include a comment that describes whatever this calculation is based upon so it's more obvious.

File: lib/improver/psychrometric_calculations/psychrometric_calculations.py (hunk 22 above)
Comment (bayliffe) [reply lib/improver/psychrometric_calculations/psychrometric_calculations.py]: The typo was the additional "p", KilopPascals, I have no problem with your capitalisation.

**Comment by gavinevans** [comment]
Closing this pull request, as this work will now be picked up in #290 and #291, now that we've agreed an approach. Much of the work from this pull request will be carried forward into #290.
